packages/director/src/execution/mongo/runs/run.model.ts

```diff
@@ -3,15 +3,74 @@ import { getMongoDB } from '@src/lib/mongo';
 import { AppError, RUN_EXISTS, CLAIM_FAILED } from '@src/lib/errors';
 import { getSanitizedMongoObject } from '@src/lib/results';
 
+const mergeRunSpecs = (run: any) => {
+  // merge fullspec into spec
+  run.specs = run.specs.map((spec: any) =>
+    Object.assign(
+      {},
+      spec,
+      run.specsFull.find((full: any) => full.instanceId === spec.instanceId) ||
+        {}
+    )
+  );
+  return run;
+};
+
+const matchRunAggregation = (runId: string) => ({
+  $match: {
+    runId,
+  },
+});
+
+const projectAggregation = {
+  $project: {
+    _id: 1,
+    runId: 1,
+    meta: 1,
+    specs: 1,
+    createdAt: 1,
+    specsFull: {
+      $map: {
+        input: '$specs',
+        as: 'spec',
+        in: '$$spec.instanceId',
+      },
+    },
+  },
+};
+
+const lookupAggregation = {
+  $lookup: {
+    from: 'instances',
+    localField: 'specsFull',
+    foreignField: 'instanceId',
+    as: 'specsFull',
+  },
+};
+
+export const getRunWithSpecs = async (id: string) =>
+  mergeRunSpecs(
+    (
+      await getMongoDB()
+        .collection('runs')
+        .aggregate([
+          matchRunAggregation(id),
+          projectAggregation,
+          lookupAggregation,
+        ])
+        .toArray()
+    ).pop()
+  );
```

Comment:
looking at this probably we'll need a proper solution for code sharing, no action required now though

Reply:
Yup I actually meant to comment about this. I attempted lots of stuff. I tried to copy in all the src folders into the dev and prod containers and setup symlinks to @sorry-cypress/<package-name> for all the projects. That blew up due to missing dependency chains. So I took a different approach and I copied the same structure of file hierarchy into the docker containers. I copied all the package manifest stuff first. Then ran `yarn` then tried to import dependancies. With this approach I had some weird inconsistency in the typescript configuration across <packages>. I tried one more thing but Im blanking on it right now. After wasting a couple days on it I backed off and did this. But I think a better solution may be to just have a separate <package> that holds all the shared files and have all the other <pacakges> use it as a dependency in thier package.json's.


I learned the node docker images have yarn. Who knew?

Reply:
Yeah, seems like dockerizing the packages prevents using yarn workspaces properly. I will have to take a deeper look, it's an interesting problem.